# **1.1 Formulation:**
Let $x_{i,j}$ be the decision variable that ith factory is located at jth location. Then $c_{j,l}$ is the cost of moving one tonne from jth location to lth location and $q_{i,k}$ is the quantity moved from ith factory to kth factory. Let $y_{i,j,k,l} = x_{i,j} * x_{k,l}$

$min \ \sum_{i,j,k,l} \ c_{j,l}* q_{i,k} * y_{i,j,k,l} \ \forall \ i,j,k,l \in \{1,2,..,n\} \\ s.t. \sum_{i=1}^n x_{i,j} = 1 \ ∀ \ j \\ \sum_{j=1}^n x_{i,j} = 1 \ ∀ \  i \\ \sum_{i=1}^{n} y_{i,j,k,l} = x_{k,l} \ \forall \ j,k,l \\ \sum_{j=1}^{n} y_{i,j,k,l} = x_{k,l} \ \forall \ i,k,l \\ x_{i,j} \in \{0,1\} \ ∀ \ i,j \\ y_{i,j,k,l} \in \{0,1\} \ ∀ \ i,j,k,l$

When there are n locations:

Number of variables = $n^4 + n^2$ variables, $n^4$ - y variables and $n^2$ - x variables.

Number of constraints = $2(n^3 + n)$



In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
q = np.loadtxt('lab7_ex1_q.txt', delimiter = '   ')
c = np.loadtxt('lab7_ex1_c.txt', delimiter = '   ')

In [ ]:
qty_data = q[:9,:9]
cost_data = c[:9,:9]

In [ ]:
qty_data

array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.],
       [ 1.,  0., 18., 22., 19., 23., 20., 18., 19.],
       [ 2., 21.,  0., 20.,  0., 19.,  0., 22., 20.],
       [ 3., 22., 23.,  0.,  0.,  0., 17., 16., 24.],
       [ 4., 17., 25.,  0.,  0., 21., 22., 20., 17.],
       [ 5., 12., 19., 18., 19.,  0., 21., 21., 23.],
       [ 6., 20.,  0.,  0., 17., 21.,  0., 20.,  0.],
       [ 7., 22., 24., 28., 16., 23.,  0.,  0., 19.],
       [ 8., 23., 29., 20., 17., 24., 24., 23.,  0.]])

In [ ]:
cost_data

array([[ 0.,  1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.],
       [ 1.,  0., 11., 13.,  9., 19., 20., 27., 13.],
       [ 2., 11.,  0.,  8.,  9., 11., 19., 10., 15.],
       [ 3., 13.,  8.,  0., 18., 19., 19., 27., 27.],
       [ 4.,  9.,  9., 18.,  0., 19., 20., 10., 20.],
       [ 5., 19., 11., 25., 19.,  0., 21., 17., 26.],
       [ 6., 20., 13., 20., 20., 21.,  0., 28., 14.],
       [ 7., 27., 10., 17., 10., 17., 28.,  0., 12.],
       [ 8., 13., 15., 27., 20., 19., 14., 12.,  0.]])

In [ ]:
model_ex1 = ConcreteModel()

In [ ]:
#No of factories
M = qty_data.shape[0] - 1

#No of locations
N = qty_data.shape[1] - 1 

In [ ]:
row_indices1 = np.arange(M)
col_indices1 = np.arange(N)
row_indices2 = np.arange(M)
col_indices2 = np.arange(N)

In [ ]:
model_ex1.x = Var(row_indices1, row_indices1, domain = Binary)
model_ex1.y = Var(row_indices1,row_indices2,col_indices1,col_indices2, domain=Binary)

In [ ]:
model_ex1.const = ConstraintList()

In [ ]:
model_ex1.obj = Objective(expr = sum(qty_data[i+1][k+1]*cost_data[j+1][l+1]*model_ex1.y[i,j,k,l] for i in row_indices1 for j in row_indices2 for k in col_indices1 for l in col_indices2), sense = minimize)

In [ ]:
for j in row_indices2:
   model_ex1.const.add(expr = sum(model_ex1.x[i,j] for i in row_indices1) == 1)
for i in row_indices1:
   model_ex1.const.add(expr = sum(model_ex1.x[i,j] for j in row_indices2) == 1)

In [ ]:
for j in row_indices2:
  for k in col_indices1:
    for l in col_indices2:
      model_ex1.const.add(expr = sum(model_ex1.y[i,j,k,l] for i in row_indices1) - model_ex1.x[k,l] == 0)

for i in row_indices1:
  for k in col_indices1:
    for l in col_indices2:
      model_ex1.const.add(expr = sum(model_ex1.y[i,j,k,l] for j in row_indices2) - model_ex1.x[k,l] == 0)


In [ ]:
model_ex1.pprint()

Streaming output truncated to the last 5000 lines.
        (0, 2, 2, 3) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 2, 4) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 2, 5) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 2, 6) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 2, 7) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 0) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 1) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 2) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 3) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 4) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 5) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 6) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 3, 7) :     0 :  None :     1 : False :  True : Binary
        (0, 2, 4, 0) :     0 :  None : 

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_ex1)
print(result)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 10573.0
  Upper bound: 10573.0
  Number of objectives: 1
  Number of constraints: 1040
  Number of variables: 4160
  Number of binary variables: 4160
  Number of integer variables: 4160
  Number of nonzeros: 2632
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.24
  Wallclock time: 0.27
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.3018758296966553
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_ex1.obj())
for i in row_indices1:
  for j in row_indices2:
    if(model_ex1.x[i,j].value != 0):
      print('{}th factory is located on the {}th location'.format(i+1,j+1))


Objective =  10573.0
1th factory is located on the 5th location
2th factory is located on the 4th location
3th factory is located on the 8th location
4th factory is located on the 6th location
5th factory is located on the 1th location
6th factory is located on the 7th location
7th factory is located on the 2th location
8th factory is located on the 3th location


## **Answer 1.9:**
**Objective Function Value** = 10573

**Allocation of factory to location:**

1 $→$ 2

2 $→$ 8

3 $→$ 7

4 $→$ 6

5 $→$ 1

6 $→$ 3

7 $→$ 5

8 $→$ 4

In [ ]:
model_ex1.x.domain=NonNegativeReals
model_ex1.y.domain=NonNegativeReals

In [ ]:
model_ex1.x.setub(1)
model_ex1.y.setub(1)

In [ ]:
result = opt_cbc.solve(model_ex1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: 10573.0
  Upper bound: 10573.0
  Number of objectives: 1
  Number of constraints: 1041
  Number of variables: 4161
  Number of nonzeros: 2632
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.13
  Wallclock time: 0.12
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 1716
  Error rc: 0
  Time: 0.1389460563659668
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_ex1.obj())
print('Decision Variables:')
for i in row_indices1:
  for j in row_indices2:
    if(model_ex1.x[i,j].value != 0):
      print('Value of Decision Variable x[',i,'][',j,']=',model_ex1.x[i,j].value)


Objective =  10573.0
Decision Variables:
Value of Decision Variable x[ 0 ][ 4 ]= 1.0
Value of Decision Variable x[ 1 ][ 3 ]= 1.0
Value of Decision Variable x[ 2 ][ 7 ]= 1.0
Value of Decision Variable x[ 3 ][ 5 ]= 1.0
Value of Decision Variable x[ 4 ][ 0 ]= 1.0
Value of Decision Variable x[ 5 ][ 6 ]= 1.0
Value of Decision Variable x[ 6 ][ 1 ]= 1.0
Value of Decision Variable x[ 7 ][ 2 ]= 1.0


## **Answer 1.10:**
**Objective Function Value** = 10573

**Decision Variables:**

x[ 0 ][ 4 ]= 1

x[ 1 ][ 3 ]= 1

x[ 2 ][ 7 ]= 1

x[ 3 ][ 5 ]= 1

x[ 4 ][ 0 ]= 1

x[ 5 ][ 6 ]= 1

x[ 6 ][ 1 ]= 1

x[ 7 ][ 2 ]= 1

## **Answer 1.11**

Yes, the optimal cost in both the cases comes out to be same and the values of the decision variables are integers as well.This is because here we are taking a convex combination of costs along the rows as well as the columns.(because the sum of the coefficients of the costs sums to 1).And we know convex combination of n numbers is greater than or equal to the minimum of those n values.So total cost is minimized when the value of the convex combination of the row/column is equal to the minimum cost along the row/column.And the minimum value is attained when the coefficient of the least value is 1 and all other coefficients are 0.